### PR TITLE
API: added stream as coll argument

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**

### CODE REVIEW

File: src/api/xccl.h [hunk 1]
```diff
@@ -435,17 +436,27 @@ typedef struct xccl_coll_algorithm {
     uint8_t id : 7;
 } xccl_coll_algorithm_t;
 
+typedef struct xccl_stream {
+    ucs_memory_type_t mem_type;
```
Comment (vspetrov) [inline src/api/xccl.h]: Why "mem_type" is part of stream ? Is it an identifier of stream "type", like CUDA_STREAM, ROCM_STREAM ... ? Maybe need to define XCCL_STREAM_TYPE?

File: src/team_lib/ucx/alltoall/alltoall_pairwise.c [hunk 2]
```diff
@@ -256,5 +262,11 @@ xccl_status_t xccl_ucx_alltoall_pairwise_start(xccl_ucx_collreq_t *req)
 
     req->alltoall_pairwise.n_rreqs       = total_reqs;
     req->alltoall_pairwise.n_sreqs       = total_reqs;
+
+    if (req->args.field_mask & XCCL_COLL_OP_ARGS_FIELD_STREAM) {
+        assert(req->src_mem_type != UCS_MEMORY_TYPE_HOST);
```
Comment (vspetrov) [inline src/team_lib/ucx/alltoall/alltoall_pairwise.c]: Why such assert? Can't i run collective over HOST buffers but have a cuda stream in my application, and use stream argument ?

File: src/team_lib/ucx/xccl_ucx_lib.h [hunk 3]
```diff
@@ -69,6 +70,7 @@ typedef struct xccl_ucx_collreq {
     xccl_tl_team_t      *team;
     xccl_status_t       complete;
     uint16_t            tag;
+    xccl_mem_component_stream_request_t *stream_req;
```
Comment (vspetrov) [inline src/team_lib/ucx/xccl_ucx_lib.h:73]: align?

File: src/utils/cuda/cuda_mem_component.c [hunk 4]
```diff
@@ -27,16 +29,59 @@ static xccl_status_t xccl_cuda_mem_free(void *ptr)
...
+{
+    int i;
+
+    CUDACHECK(cudaStreamCreateWithFlags(&xccl_cuda_mem_component.stream,
+                                        cudaStreamNonBlocking));
+    CUDACHECK(cudaHostAlloc((void**)&xccl_cuda_mem_component.stream_requests,
+              NUM_STREAM_REQUESTS * sizeof(xccl_cuda_mem_component_stream_request_t),
+              cudaHostAllocMapped));
+    for (i = 0; i < NUM_STREAM_REQUESTS; i++) {
+        xccl_cuda_mem_component.stream_requests[i].is_free = 1;
```
Comment (vspetrov) [inline src/utils/cuda/cuda_mem_component.c:52]: this looks like a pool of stream_reqs - is it thread safe (does not look like ) ? Can we use ucs mpool for this ?

File: src/utils/cuda/cuda_mem_component.c [hunk 5]
```diff
@@ -27,16 +29,59 @@ static xccl_status_t xccl_cuda_mem_free(void *ptr)
...
 xccl_status_t xccl_cuda_reduce_impl(void *sbuf1, void *sbuf2, void *target,
                                     size_t count, xccl_dt_t dtype, xccl_op_t op,
                                     cudaStream_t stream);
 
 xccl_status_t xccl_cuda_reduce(void *sbuf1, void *sbuf2, void *target,
                                size_t count, xccl_dt_t dtype, xccl_op_t op)
 {
+    xccl_status_t st;
+
     if (xccl_cuda_mem_component.stream == 0) {
```
Comment (vspetrov) [inline src/utils/cuda/cuda_mem_component.c]: Maybe use 1-line macro for the whole this if () {.. } block - it is used multiple times

**Review by vspetrov — CHANGES_REQUESTED** [review]

File: src/team_lib/ucx/alltoall/alltoall_pairwise.c (hunk 2 above)
Comment (Sergei-Lebedev) [reply src/team_lib/ucx/alltoall/alltoall_pairwise.c]: Do you mean use case where we start collective with host buffers only after some gpu activity done?

File: src/utils/cuda/cuda_mem_component.c (hunk 4 above)
Comment (Sergei-Lebedev) [reply src/utils/cuda/cuda_mem_component.c:52]: No, it's not thread safe, we need some global lock to access it. And I like the idea of using mpool, should work here, but let's make it in next PR?

File: src/team_lib/ucx/alltoall/alltoall_pairwise.c (hunk 2 above)
Comment (vspetrov) [reply src/team_lib/ucx/alltoall/alltoall_pairwise.c]: yes, of vice-versa start collective over host buffer after gpu kernel is done. I know sounds weird and can be resolved at user level. So just asking if we "require" from the user such assert ?

File: src/utils/cuda/cuda_mem_component.c (hunk 4 above)
Comment (vspetrov) [reply src/utils/cuda/cuda_mem_component.c:52]: sure 👍

**Review by vspetrov — APPROVED** [review]

File: src/team_lib/ucx/alltoall/alltoall_pairwise.c (hunk 2 above)
Comment (Sergei-Lebedev) [reply src/team_lib/ucx/alltoall/alltoall_pairwise.c]: you are right, will remove assert

**Pull request merged**